dlib/dnn/loss.h

```diff
@@ -3447,6 +3447,488 @@ namespace dlib
+                        {
+                            loss += -scale_loss * options.lambda_cls * safe_log(out_data[c_idx]);
+                            g[c_idx] = scale_grad * options.lambda_cls * (out_data[c_idx] - 1);
+                        }
+                        else
+                        {
+                            loss += -scale_loss * options.lambda_cls * safe_log(1 - out_data[c_idx]);
+                            g[c_idx] = scale_grad * options.lambda_cls * out_data[c_idx];
+                        }
+                    }
+                }
+            }
+        };
+
+    }  // namespace impl
+
+    template <template <typename> class... TAG_TYPES>
+    class loss_yolo_
+    {
+        static void list_tags(std::ostream& out) { impl::yolo_helper_impl<TAG_TYPES...>::list_tags(out); }
+
+    public:
+
+        typedef std::vector<yolo_rect> training_label_type;
+        typedef std::vector<yolo_rect> output_label_type;
+
+        constexpr static size_t tag_count() { return impl::yolo_helper_impl<TAG_TYPES...>::tag_count(); }
+
+        loss_yolo_() {};
+
+        loss_yolo_(const yolo_options& options) : options(options) { }
+
+        template <
+            typename SUB_TYPE,
+            typename label_iterator
+            >
+        void to_label (
+            const tensor& input_tensor,
+            const SUB_TYPE& sub,
+            label_iterator iter
+        ) const
+        {
+            DLIB_CASSERT(sub.sample_expansion_factor() == 1, sub.sample_expansion_factor());
+            std::vector<yolo_rect> dets_accum;
+            output_label_type final_dets;
+            for (long i = 0; i < input_tensor.num_samples(); ++i)
+            {
+                dets_accum.clear();
+                impl::yolo_helper_impl<TAG_TYPES...>::tensor_to_dets(input_tensor, sub, i, options, dets_accum);
+
+                // Do non-max suppression
+                std::sort(dets_accum.rbegin(), dets_accum.rend());
+                final_dets.clear();
+                for (size_t j = 0; j < dets_accum.size(); ++j)
+                {
+                    if (overlaps_any_box_nms(final_dets, dets_accum[j]))
+                        continue;
+
+                    final_dets.push_back(dets_accum[j]);
+                }
+
+                *iter++ = std::move(final_dets);
+            }
+        }
+
+        template <
+            typename const_label_iterator,
+            typename SUBNET
+        >
+        double compute_loss_value_and_gradient (
+            const tensor& input_tensor,
+            const_label_iterator truth,
+            SUBNET& sub
+        ) const
+        {
+            DLIB_CASSERT(input_tensor.num_samples() != 0);
+            DLIB_CASSERT(sub.sample_expansion_factor() == 1);
+            double loss = 0;
+            for (long i = 0; i < input_tensor.num_samples(); ++i)
+            {
+                impl::yolo_helper_impl<TAG_TYPES...>::tensor_to_grad(input_tensor, truth, sub, i, options, loss);
```

Comment:
One doubt I have comes from here.

When `back_propagate_error` will be called, I've seen that it will clear the gradient inputs. I am worried that this is happening and even if I fill the gradients for the 3 outputs, the gradients that are not at the top get zeroed out at backpropagation.

That would explain why it's not learning (but it would make it harder to fix...) I hope I just made a mistake in my implementation.